torch_geometric/utils/assortativity.py

```diff
@@ -0,0 +1,65 @@
+import torch
+from torch_sparse import SparseTensor
+
+from torch_geometric.typing import Adj
+from torch_geometric.utils import coalesce, degree, to_dense_adj
+
+
+def assortativity(edge_index: Adj) -> float:
+    r"""The degree assortativity coeffeicent from the
+    `"Mixing patterns in networks"
+    <https://arxiv.org/abs/cond-mat/0209450>`_ paper
+    Assortativity measures the similarity of connections in
+    the graph with respect to the node degree. It is computed by
+    Pearson correlation coefficient of the node degrees.
+
+    Args:
+        edge_index (Tensor or SparseTensor): The graph connectivity.
+
+    Returns:
+        The value of the degree assortativity coefficient for the input
+        graph :math:`\in [-1, 1]`
+
+    Example:
+
+        >>> edge_index = torch.tensor([[0, 1, 2, 3, 2],
+        ...                            [1, 2, 0, 1, 3]])
+        >>> assortativity(edge_index)
+        -0.666667640209198
+    """
+    if isinstance(edge_index, SparseTensor):
+        row, col, _ = edge_index.coo()
+    else:
+        row, col = edge_index
+
+    device = row.device
+    out_deg = degree(row, dtype=torch.long)
+    in_deg = degree(col, dtype=torch.long)
+    degrees = torch.unique(torch.cat([out_deg, in_deg]))
+    mapping = row.new_zeros(degrees.max().item() + 1)
+    mapping[degrees] = torch.arange(degrees.size(0), device=device)
+
+    # Compute degree mixing matrix (joint probability distribution) `M`
+    num_degrees = degrees.size(0)
+    M = torch.zeros(num_degrees, num_degrees, device=device)
```

Comment:
This matrix seems to not be used.

Reply:
Used later to compute row/col sum
```python
a, b = M.sum(0), M.sum(1)
```

Reply:
Not sure I understand. You re-assign it in line 51.

Reply:
Oh I missed it. Now it has been fixed (removed).